### PR TITLE
Uplift `clippy::undropped_manually_drops` lint

### PR DESCRIPTION
**Pull request by Urgau**
This PR aims at uplifting the `clippy::undropped_manually_drops` lint.

## `undropped_manually_drops`

(deny-by-default)

The `undropped_manually_drops` lint check for calls to `std::mem::drop` with a value of `std::mem::ManuallyDrop` which doesn't drop.

### Example

```rust
struct S;
drop(std::mem::ManuallyDrop::new(S));
```

### Explanation

`ManuallyDrop` does not drop it's inner value so calling `std::mem::drop` will not drop the inner value of the `ManuallyDrop` either.

-----

Mostly followed the instructions for uplifting an clippy lint described here: https://github.com/rust-lang/rust/pull/99696#pullrequestreview-1134072751

@rustbot label: +I-lang-nominated
r? compiler

-----

For Clippy:

changelog: Moves: Uplifted `clippy::undropped_manually_drops` into rustc

### CODE REVIEW

**Comment by asquared31415** [comment]
> ManuallyDrop doesn't not drop it's inner value

I think this part of the description is incorrect? Should it be "`ManuallyDrop` *does not* drop *its* inner value ..."?

**Comment by tmandry** [comment]
@rfcbot fcp merge

**Comment by tmandry** [comment]
@rfcbot fcp merge

**Comment by scottmcm** [comment]
Didn't I see some lints go by recently that lint for using `drop` on things that are `Copy`?  Might it make sense for this to be just part of that lint?  Something like "promised it won't be `needs_drop`" instead of `Copy`?

Or if this is worth adding separately, maybe it would make sense to give it a higher default level as a "this is a specific mistake that people make" deny lint rather than a "this is useless" warn lint?

**Comment by scottmcm** [comment]
Yeah, we have a couple of other deny-by-default lints like that.  For example,
- unused binding from match pattern: warning
- unused binding from match pattern, but spelled exactly the same as a variant of the type it's matching against: deny (and suggest replacing `A => ` with `Foo::A => `)

When there's a super-specific suggestion we can give (here to use `ManuallyDrop::drop` instead of `mem::drop`) because it's a known pattern, I think `deny`-by-default makes good sense.

**Comment by Urgau** [comment]
> Yeah, we have a couple of other deny-by-default lints like that.
> 
>   [..]
> 
> When there's a super-specific suggestion we can give (here to use `ManuallyDrop::drop` instead of `mem::drop`) because it's a known pattern, I think `deny`-by-default makes good sense.

Make sense, I will change it to deny-by-default.

(btw, the suggestion here is `ManuallyDrop::into_inner` not `ManuallyDrop::drop` as the lint only triggers on owned `ManuallyDrop`, the ref cases and handled by the other lints)

**Comment by Urgau** [comment]
r? compiler

File: compiler/rustc_lint/src/drop_forget_useless.rs [hunk 1]
```diff
@@ -162,3 +202,10 @@ fn is_single_call_in_arm<'tcx>(
     }
     false
 }
+
+fn is_type_manually_drop(cx: &LateContext<'_>, ty: Ty<'_>) -> bool {
+    match ty.kind() {
+        ty::Adt(adt, _) => cx.tcx.lang_items().get(LangItem::ManuallyDrop) == Some(adt.did()),
```
Comment (compiler-errors) [inline compiler/rustc_lint/src/drop_forget_useless.rs:208]: We already have that info stored in the ADT flags: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/adt/struct.AdtDef.html#method.is_manually_drop

**Review by compiler-errors — COMMENTED** [review]
one nit, r=me after

**Comment by Urgau** [comment]
> (delegate still works even though bors didn't comment, just use `@bors` `r=compiler-errors` when this is ready to go)

unfortunately it won't work since you have used `@rustbot` and not `@bors` (unless I'm missing something)

@rustbot ready

**Comment by GuillaumeGomez** [comment]
@compiler-errors I think it was supposed to be done using `bors` and not `rustbot`. ;)

@bors r=compiler-errors

**Comment by compiler-errors** [comment]
🤦

Whoops, I mixed "compiler-errors makes a typo" up with the actual "bors delegate doesn't tell you the delegate succeeded" bug (https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/bors.20ignoring.20.60delegate.2B.60.3F) -- thx @GuillaumeGomez

**Comment by pnkfelix** [comment]
* only primary benchmark to regress was helloworld (3 incr check variants), and not by all that much (relatively speaking)
* secondary regressions were solely to unify-linearly, await-call-tree, token-stream-stress.
* impact seems acceptable (one expects new lint to add some amount of extra work, and I wouldn't be surprised if this is actually noise,. given #112116 above).
* marking as triaged.

@rustbot label: +perf-regression-triaged

**Pull request merged**